lib/cosmos/streams/length_stream_protocol.rb

```diff
@@ -75,41 +75,64 @@ def initialize(
       @max_length = Integer(@max_length) if @max_length
     end
 
-    # See StreamProtocol#pre_write_packet
+    # Called to perform modifications on a command packet before it is send
+    #
+    # @param packet [Packet] Original packet
+    # @return [Packet] Potentially modified packet
     def pre_write_packet(packet)
-      data = super(packet)
-      if @fill_sync_pattern # and length
-        # Fill the length field
-        length = (data.length - @length_value_offset) / @length_bytes_per_count
-        BinaryAccessor.write(length,
-          @length_bit_offset,
-          @length_bit_size,
-          :UINT,
-          data,
-          @length_endianness,
-          :ERROR)
-
-        # Also write the new length field into the packet that will be logged if it exists in the packet
-        if @discard_leading_bytes > 0
-          # The write above did not write into the original packet
-          original_length_bit_offset = @length_bit_offset - (@discard_leading_bytes * 8)
-          if original_length_bit_offset >= 0
-            original_data = packet.buffer(false)
-            BinaryAccessor.write(length,
-              original_length_bit_offset,
-              @length_bit_size,
-              :UINT,
-              original_data,
-              @length_endianness,
-              :ERROR)
-          end
+      packet = super(packet) # Allow stream_protocol to set the sync if needed
```

Comment:
I think the 'super' for pre_write_packet should always be first so you ensure they don't write over what you're doing.